Ftp Connector: retrieveFromPath does not return all bytes of file

Ftp connector is able to retrieve the full contents of small file, e.g. 30020 bytes, but it does not for a larger file, e.g. 2000020 bytes.

@viktorklang I've added a test in 	rafacm/alpakka@e826604 to reproduce this issue and currently outputs the following error and makes the test assertion to time out:
```
[ERROR] [01/20/2017 11:34:52.420] [default-akka.stream.default-blocking-io-dispatcher-13] [akka://default/user/StreamSupervisor-0/flow-3-0-FtpIOSource] Error during postStop in [akka.stream.alpakka.ftp.impl.FtpSourceFactory$$anon$2@594ce681]: Connection closed without indication.
org.apache.commons.net.ftp.FTPConnectionClosedException: Connection closed without indication.
        at org.apache.commons.net.ftp.FTP.__getReply(FTP.java:316)
        at org.apache.commons.net.ftp.FTP.__getReply(FTP.java:292)
        at org.apache.commons.net.ftp.FTP.sendCommand(FTP.java:503)
        at org.apache.commons.net.ftp.FTP.sendCommand(FTP.java:628)
        at org.apache.commons.net.ftp.FTP.sendCommand(FTP.java:602)
        at org.apache.commons.net.ftp.FTP.quit(FTP.java:884)
        at org.apache.commons.net.ftp.FTPClient.logout(FTPClient.java:1152)
        at akka.stream.alpakka.ftp.impl.FtpOperations$class.disconnect(FtpOperations.scala:34)
        at akka.stream.alpakka.ftp.impl.FtpLike$$anon$1.disconnect(FtpLike.scala:31)
        at akka.stream.alpakka.ftp.impl.FtpLike$$anon$1.disconnect(FtpLike.scala:31)
        at akka.stream.alpakka.ftp.impl.FtpGraphStageLogic.disconnect(FtpGraphStageLogic.scala:51)
        at akka.stream.alpakka.ftp.impl.FtpGraphStageLogic.postStop(FtpGraphStageLogic.scala:42)
        at akka.stream.alpakka.ftp.impl.FtpIOGraphStage$$anon$1.postStop(FtpIOGraphStage.scala:76)
        at akka.stream.impl.fusing.GraphInterpreter.finalizeStage(GraphInterpreter.scala:801)
        at akka.stream.impl.fusing.GraphInterpreter.finish(GraphInterpreter.scala:538)
        at akka.stream.impl.fusing.GraphInterpreterShell.tryAbort(ActorGraphInterpreter.scala:509)
        at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$postStop$1.apply(ActorGraphInterpreter.scala:641)
        at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$postStop$1.apply(ActorGraphInterpreter.scala:641)
        at scala.collection.immutable.Set$Set1.foreach(Set.scala:94)
        at akka.stream.impl.fusing.ActorGraphInterpreter.postStop(ActorGraphInterpreter.scala:641)
        at akka.actor.Actor$class.aroundPostStop(Actor.scala:509)
        at akka.stream.impl.fusing.ActorGraphInterpreter.aroundPostStop(ActorGraphInterpreter.scala:529)
        at akka.actor.dungeon.FaultHandling$class.akka$actor$dungeon$FaultHandling$$finishTerminate(FaultHandling.scala:210)
        at akka.actor.dungeon.FaultHandling$class.terminate(FaultHandling.scala:172)
        at akka.actor.ActorCell.terminate(ActorCell.scala:374)
        at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:467)
        at akka.actor.ActorCell.systemInvoke(ActorCell.scala:483)
        at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
        at akka.dispatch.Mailbox.run(Mailbox.scala:223)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

[info] FtpSourceSpec:
[info] FtpBrowserSource
[info] - should list all files from root (191 milliseconds)
[info] - should list all files from non-root (32 milliseconds)
[info] FtpIOSource
[info] - should retrieve a file from path as a stream of bytes (32 milliseconds)
[info] - should retrieve a bigger file (~2 MB) from path as a stream of bytes (reproduces issue #154) *** FAILED *** (179 milliseconds)
[info]   A timeout occurred waiting for a future to complete. Queried 10 times, sleeping 15 milliseconds between each query. (CommonFtpSourceSpec.scala:64)
[info]   org.scalatest.concurrent.Futures$FutureConcept$$anon$1:
[info]   at org.scalatest.concurrent.Futures$FutureConcept$class.tryTryAgain$1(Futures.scala:538)
[info]   at org.scalatest.concurrent.Futures$FutureConcept$class.futureValueImpl(Futures.scala:550)
[info]   at org.scalatest.concurrent.ScalaFutures$$anon$1.futureValueImpl(ScalaFutures.scala:275)
[info]   at org.scalatest.concurrent.Futures$FutureConcept$class.futureValue(Futures.scala:476)
[info]   at org.scalatest.concurrent.ScalaFutures$$anon$1.futureValue(ScalaFutures.scala:275)
[info]   at akka.stream.alpakka.ftp.CommonFtpSourceSpec$$anonfun$2$$anonfun$apply$mcV$sp$4.apply(CommonFtpSourceSpec.scala:64)
[info]   at akka.stream.alpakka.ftp.CommonFtpSourceSpec$$anonfun$2$$anonfun$apply$mcV$sp$4.apply(CommonFtpSourceSpec.scala:55)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.WordSpecLike$$anon$1.apply(WordSpecLike.scala:1078)
[info]   at org.scalatest.TestSuite$class.withFixture(TestSuite.scala:196)
[info]   at akka.stream.alpakka.ftp.FtpSourceSpec.withFixture(CommonFtpSourceSpec.scala:12)
[info]   at org.scalatest.WordSpecLike$class.invokeWithFixture$1(WordSpecLike.scala:1075)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:1088)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:1088)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
[info]   at org.scalatest.WordSpecLike$class.runTest(WordSpecLike.scala:1088)
[info]   at akka.stream.alpakka.ftp.FtpSourceSpec.org$scalatest$BeforeAndAfter$$super$runTest(CommonFtpSourceSpec.scala:12)
[info]   at org.scalatest.BeforeAndAfter$class.runTest(BeforeAndAfter.scala:203)
[info]   at akka.stream.alpakka.ftp.FtpSourceSpec.runTest(CommonFtpSourceSpec.scala:12)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1147)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1147)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:396)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
[info]   at scala.collection.immutable.List.foreach(List.scala:381)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:373)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:410)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
[info]   at scala.collection.immutable.List.foreach(List.scala:381)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:379)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
[info]   at org.scalatest.WordSpecLike$class.runTests(WordSpecLike.scala:1147)
[info]   at akka.stream.alpakka.ftp.FtpSourceSpec.runTests(CommonFtpSourceSpec.scala:12)
[info]   at org.scalatest.Suite$class.run(Suite.scala:1147)
[info]   at akka.stream.alpakka.ftp.FtpSourceSpec.org$scalatest$WordSpecLike$$super$run(CommonFtpSourceSpec.scala:12)
[info]   at org.scalatest.WordSpecLike$$anonfun$run$1.apply(WordSpecLike.scala:1192)
[info]   at org.scalatest.WordSpecLike$$anonfun$run$1.apply(WordSpecLike.scala:1192)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
[info]   at org.scalatest.WordSpecLike$class.run(WordSpecLike.scala:1192)
[info]   at akka.stream.alpakka.ftp.FtpSourceSpec.org$scalatest$BeforeAndAfter$$super$run(CommonFtpSourceSpec.scala:12)
[info]   at org.scalatest.BeforeAndAfter$class.run(BeforeAndAfter.scala:258)
[info]   at akka.stream.alpakka.ftp.FtpSourceSpec.org$scalatest$BeforeAndAfterAll$$super$run(CommonFtpSourceSpec.scala:12)
[info]   at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:210)
[info]   at akka.stream.alpakka.ftp.FtpSourceSpec.run(CommonFtpSourceSpec.scala:12)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:314)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:472)
[info]   at sbt.TestRunner.runTest$1(TestFramework.scala:76)
[info]   at sbt.TestRunner.run(TestFramework.scala:85)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
[info]   at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
[info]   at sbt.TestFunction.apply(TestFramework.scala:207)
[info]   at sbt.Tests$.sbt$Tests$$processRunnable$1(Tests.scala:239)
[info]   at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
[info]   at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
[info]   at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
[info]   at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
[info]   at sbt.std.Transform$$anon$4.work(System.scala:63)
[info]   at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
[info]   at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
[info]   at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
[info]   at sbt.Execute.work(Execute.scala:237)
[info]   at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
[info]   at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
[info]   at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
[info]   at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[info]   at java.lang.Thread.run(Thread.java:745)
```

--- a/ftp/src/test/java/akka/stream/alpakka/ftp/FtpBaseSupport.java
+++ b/ftp/src/test/java/akka/stream/alpakka/ftp/FtpBaseSupport.java
@@ -96,13 +96,17 @@ abstract class FtpBaseSupport implements FtpSupport, AkkaSupport {
     }
 
     public void putFileOnFtp(String path, String fileName) {
+        putFileOnFtpWithContents(path, fileName, loremIpsum.getBytes());
+    }
+
+    public void putFileOnFtpWithContents(String path, String fileName, byte[] fileContents) {
         try {
             Path baseDir = getFileSystem().getPath(path);
             if (!Files.exists(baseDir)) {
                 Files.createDirectories(baseDir);
             }
             Path filePath = baseDir.resolve(fileName);
-            Files.write(filePath, loremIpsum.getBytes());
+            Files.write(filePath, fileContents);
         } catch (Throwable t) {
             throw new RuntimeException(t);
         }

--- a/ftp/src/test/java/akka/stream/alpakka/ftp/FtpSupport.java
+++ b/ftp/src/test/java/akka/stream/alpakka/ftp/FtpSupport.java
@@ -21,5 +21,7 @@ interface FtpSupport {
 
     void putFileOnFtp(String path, String fileName);
 
+    void putFileOnFtpWithContents(String path, String fileName, byte[] fileContents);
+
     String getLoremIpsum();
 }

--- a/ftp/src/test/scala/akka/stream/alpakka/ftp/BaseFtpSpec.scala
+++ b/ftp/src/test/scala/akka/stream/alpakka/ftp/BaseFtpSpec.scala
@@ -20,6 +20,7 @@ trait BaseFtpSpec extends PlainFtpSupportImpl with BaseSpec {
     InetAddress.getByName("localhost"),
     getPort,
     AnonFtpCredentials,
+    binary = true,
     passiveMode = true
   )
   //#create-settings

--- a/ftp/src/test/scala/akka/stream/alpakka/ftp/BaseFtpsSpec.scala
+++ b/ftp/src/test/scala/akka/stream/alpakka/ftp/BaseFtpsSpec.scala
@@ -20,6 +20,7 @@ trait BaseFtpsSpec extends FtpsSupportImpl with BaseSpec {
     InetAddress.getByName("localhost"),
     getPort,
     AnonFtpCredentials,
+    binary = true,
     passiveMode = true
   )
   //#create-settings

--- a/ftp/src/test/scala/akka/stream/alpakka/ftp/CommonFtpSourceSpec.scala
+++ b/ftp/src/test/scala/akka/stream/alpakka/ftp/CommonFtpSourceSpec.scala
@@ -6,6 +6,8 @@ package akka.stream.alpakka.ftp
 import akka.stream.IOResult
 import akka.stream.scaladsl.{Keep, Sink}
 import akka.stream.testkit.scaladsl.TestSink
+import org.scalatest.time.{Millis, Seconds, Span}
+import scala.util.Random
 
 final class FtpSourceSpec extends BaseFtpSpec with CommonFtpSourceSpec
 final class SftpSourceSpec extends BaseSftpSpec with CommonFtpSourceSpec
@@ -18,6 +20,8 @@ trait CommonFtpSourceSpec extends BaseSpec {
 
   implicit val system = getSystem
   implicit val mat = getMaterializer
+  implicit val defaultPatience =
+    PatienceConfig(timeout = Span(3, Seconds), interval = Span(300, Millis))
 
   "FtpBrowserSource" should {
     "list all files from root" in {
@@ -47,6 +51,18 @@ trait CommonFtpSourceSpec extends BaseSpec {
       probe.request(100).expectNextOrComplete()
 
       val expectedNumOfBytes = getLoremIpsum.getBytes().length
+      result.futureValue shouldBe IOResult.createSuccessful(expectedNumOfBytes)
+    }
+
+    "retrieve a bigger file (~2 MB) from path as a stream of bytes" in {
+      val fileName = "sample_bigger_file"
+      val fileContents = new Array[Byte](2000020)
+      Random.nextBytes(fileContents)
+      putFileOnFtpWithContents(FtpBaseSupport.FTP_ROOT_DIR, fileName, fileContents)
+      val (result, probe) = retrieveFromPath(s"/$fileName").toMat(TestSink.probe)(Keep.both).run()
+      probe.request(1000).expectNextOrComplete()
+
+      val expectedNumOfBytes = fileContents.length
       result.futureValue shouldBe IOResult.createSuccessful(expectedNumOfBytes)
     }
   }

